Engine/Rendering/Viewport.cpp

```diff
@@ -0,0 +1,536 @@
+	num_rendered_vertices = 0;
+
+	float3 camera_position = camera->owner->transform.GetGlobalTranslation();
+
+	std::vector<Utils::MeshRendererDistancePair> opaque_mesh_renderers;
+	std::vector<Utils::MeshRendererDistancePair> transparent_mesh_renderers;
+	Utils::SplitCulledMeshRenderers(culled_mesh_renderers, camera_position, opaque_mesh_renderers, transparent_mesh_renderers);
+
+	for (auto& opaque_mesh_renderer : opaque_mesh_renderers)
+	{
+		if (opaque_mesh_renderer.mesh_renderer->mesh_to_render != nullptr 
+			&& opaque_mesh_renderer.mesh_renderer->material_to_render != nullptr
+			&& opaque_mesh_renderer.mesh_renderer->IsEnabled()
+		)
+		{
+			GLuint mesh_renderer_program = opaque_mesh_renderer.mesh_renderer->BindShaderProgram();
+			opaque_mesh_renderer.mesh_renderer->BindMeshUniforms(mesh_renderer_program);
+			opaque_mesh_renderer.mesh_renderer->BindMaterialUniforms(mesh_renderer_program);
+			BindDepthMaps(mesh_renderer_program);
+			App->lights->Render(opaque_mesh_renderer.mesh_renderer->owner->transform.GetGlobalTranslation(), mesh_renderer_program);
+			opaque_mesh_renderer.mesh_renderer->RenderModel();
+
+			num_rendered_triangles += opaque_mesh_renderer.mesh_renderer->mesh_to_render->GetNumTriangles();
+			num_rendered_vertices += opaque_mesh_renderer.mesh_renderer->mesh_to_render->GetNumVerts();
+
+			glUseProgram(0);
+		}
+	}
+
+	glEnable(GL_BLEND);
+	glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+	glBlendEquation(GL_FUNC_ADD);
+	for (auto& transparent_mesh_renderer : transparent_mesh_renderers)
+	{
+		if (transparent_mesh_renderer.mesh_renderer->mesh_to_render != nullptr
+			&& transparent_mesh_renderer.mesh_renderer->material_to_render != nullptr
+			&& transparent_mesh_renderer.mesh_renderer->IsEnabled()
+		) {
+			GLuint mesh_renderer_program = transparent_mesh_renderer.mesh_renderer->BindShaderProgram();
+			transparent_mesh_renderer.mesh_renderer->BindMeshUniforms(mesh_renderer_program);
+			transparent_mesh_renderer.mesh_renderer->BindMaterialUniforms(mesh_renderer_program);
+			BindDepthMaps(mesh_renderer_program);
+			App->lights->Render(transparent_mesh_renderer.mesh_renderer->owner->transform.GetGlobalTranslation(), mesh_renderer_program);
+			transparent_mesh_renderer.mesh_renderer->RenderModel();
+
+			num_rendered_triangles += transparent_mesh_renderer.mesh_renderer->mesh_to_render->GetNumTriangles();
+			num_rendered_vertices += transparent_mesh_renderer.mesh_renderer->mesh_to_render->GetNumVerts();
+
+			glUseProgram(0);
+		}
+	}
+
+	FrameBuffer::UnBind();
+}
+
+void Viewport::EffectsRenderPass() const
+{
+	if (!effects_pass)
+	{
+		return;
+	}
+
+	scene_fbo->Bind();
+	App->effects->Render();
+	FrameBuffer::UnBind();
+}
+
+void Viewport::UIRenderPass() const
+{
+	if (!IsOptionSet(ViewportOption::RENDER_UI))
+	{
+		return;
+	}
+
+	postprocess_fbo->Bind();
+	App->ui->Render(width, height, IsOptionSet(ViewportOption::SCENE_MODE));
+	FrameBuffer::UnBind();
+}
+
+void Viewport::PostProcessPass()
```

Comment:
Can this be const?

Reply:
No, because it calls BloomPass that modifies `ping_pong_buffer`